src/v/security/oidc_authenticator.cc

```diff
@@ -0,0 +1,166 @@
+        return os << "init";
+    case state::complete:
+        return os << "complete";
+    case state::failed:
+        return os << "failed";
+    }
+}
+
+// impl runs on the thread_worker
+class oidc_authenticator::impl {
+public:
+    impl() = default;
+
+    ss::future<result<void>> authenticate(bytes auth_bytes);
+    const security::acl_principal& principal() const { return _principal; }
+
+private:
+    ss::future<result<void>> extract_principal(bytes auth_bytes);
+    security::acl_principal _principal;
+};
+
+oidc_authenticator::oidc_authenticator()
+  : _impl{std::make_unique<impl>()} {}
+
+oidc_authenticator::~oidc_authenticator() = default;
+
+ss::future<result<bytes>> oidc_authenticator::authenticate(bytes auth_bytes) {
+    if (!_impl) {
+        vlog(
+          seclog.error,
+          "authenticate received after handshake complete {} bytes {}",
+          _state,
+          auth_bytes.size());
+        co_return errc::invalid_oidc_state;
+    }
+
+    auto res = co_await _impl->authenticate(std::move(auth_bytes));
+    if (res.has_error()) {
+        _state = state::failed;
+        co_return res.assume_error();
+    }
+    _principal = _impl->principal();
+    _state = state::complete;
+    co_return bytes{};
+}
+
+ss::future<result<void>> oidc_authenticator::impl::authenticate(bytes b) {
+    vlog(seclog.trace, "oidc authenticate received {} bytes", b.size());
+    return extract_principal(std::move(b));
+}
+
+ss::future<result<void>>
+oidc_authenticator::impl::extract_principal(bytes auth_bytes) {
+    auto auth_str = std::string_view(
+      reinterpret_cast<char*>(auth_bytes.data()), auth_bytes.size());
+
+    constexpr std::string_view sasl_header{"n,,\1auth=Bearer "};
+    if (!auth_str.starts_with(sasl_header)) {
+        vlog(seclog.debug, "invalid sasl_header");
+        co_return errc::invalid_credentials;
+    }
+    auth_str = auth_str.substr(sasl_header.length());
+    if (!auth_str.ends_with("\1\1")) {
+        vlog(seclog.debug, "invalid sasl_header");
+        co_return errc::invalid_credentials;
+    }
+    auth_str = auth_str.substr(0, auth_str.length() - 2);
+    vlog(seclog.trace, "auth_bytes: {}", auth_str);
+
+    std::vector<std::string_view> jose_enc;
+    jose_enc.reserve(3);
+    boost::algorithm::split(
+      jose_enc, auth_str, [](char c) { return c == '.'; });
+    vlog(seclog.trace, "jose_b64: {}", jose_enc);
+
+    if (jose_enc.size() != 3) {
+        vlog(seclog.debug, "invalid secured JWT");
+        co_return errc::invalid_credentials;
+    }
+
+    auto base64_url_decode = [](std::string_view sv) {
+        CryptoPP::Base64URLDecoder decoder;
+
+        decoder.Put((CryptoPP::byte*)sv.data(), sv.size());
+        decoder.MessageEnd();
+        std::string decoded;
+        CryptoPP::word64 size = decoder.MaxRetrievable();
+        if (size && size <= SIZE_MAX) {
+            decoded.resize(size);
+            decoder.Get((CryptoPP::byte*)decoded.data(), decoded.size());
```

Comment:
Can any of cryptopp stuff throw? If so, would that just be handled in an enclosing scope?

Reply:
> Can any of cryptopp stuff throw? If so, would that just be handled in an enclosing scope?

Good spot, yes it can, it should get converted at the coroutine boundary. But it might not be handled well.

I plan to move all of this to a mini JWT library in the next PR where I'll clean it up.